include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+  }
+
+private:
+  // Kind must be TypeKind::Typedef, Underlying type must verify, and
+  // Underlying type must not refer directly to this type
+  bool verifyDerived() const debug_function {
+    if (getKind() != TypeKind::Typedef)
+      return false;
+
+    if (not UnderlyingType.ThePath.Root)
+      return false;
+
+    flatc::Type *Underlying = UnderlyingType.ThePath.get();
+    if (not Underlying or Underlying == this)
+      return false;
+
+    return Underlying->verify();
+  }
+
+  friend struct flatc::Type;
+};
+INTROSPECTION_NS(flatc, TypedefType, UUID, Name, Kind, UnderlyingType);
+
+/// \brief A field of an aggregate type in flatc, with qualified type and name
+struct flatc::AggregateField {
+  QualifiedType FieldType;
+  std::string Name;
+
+  AggregateField(const QualifiedType &QT, llvm::StringRef NameRef) :
+    FieldType(QT), Name(NameRef) {}
+
+  AggregateField(QualifiedType &&QT, llvm::StringRef NameRef) :
+    FieldType(std::move(QT)), Name(NameRef) {}
+
+  AggregateField(const QualifiedType &QT) : AggregateField(QT, "") {}
+
+  AggregateField(QualifiedType &&QT) : AggregateField(std::move(QT), "") {}
+
+  AggregateField() :
+    AggregateField(QualifiedType{ Path::fromString("/"), {} }) {}
+
+  ~AggregateField() = default;
+
+  AggregateField(const AggregateField &) = default;
+  AggregateField &operator=(const AggregateField &) = default;
+
+  AggregateField(AggregateField &&) = default;
+  AggregateField &operator=(AggregateField &&) = default;
+
+  bool operator==(const AggregateField &) const = default;
+
+public:
+  bool verify() const { return not Name.empty() and FieldType.verify(); }
+};
+
+/// \brief A field of a struct type in flatc, with offset, qualified type, and
+/// name
+struct flatc::StructField : flatc::AggregateField {
+  std::size_t Offset;
+
+  StructField(std::size_t O, llvm::StringRef NameRef, const QualifiedType &QT) :
+    AggregateField(QT, NameRef), Offset(O) {}
+
+  StructField(std::size_t O) : StructField(O, "", {}) {}
+
+  StructField() : StructField(0) {}
+
+  ~StructField() = default;
+
+  StructField(const StructField &) = default;
+  StructField &operator=(const StructField &) = default;
+
+  StructField(StructField &&) = default;
+  StructField &operator=(StructField &&) = default;
+
+  bool operator==(const StructField &) const = default;
+};
+INTROSPECTION_NS(flatc, StructField, FieldType, Name, Offset);
+
+/// \brief KeyedObjectTraits for flatc::StructType based on its byte-offset in
+/// struct
+template<>
+struct KeyedObjectTraits<flatc::StructField> {
+
+  static std::size_t key(const flatc::StructField &Val) { return Val.Offset; }
+
+  static flatc::StructField fromKey(const std::size_t &Offset) {
+    return flatc::StructField(Offset);
+  }
+};
+
+/// \brief A struct type in flatc. Structs are actually typedefs of unnamed
+/// structs in C.
+struct flatc::StructType : public flatc::Type {
+  static constexpr const char *Tag = "!flatc::StructType";
```

Comment:
```suggestion
  static constexpr const char *Tag = "!Struct";
```

Please use the name of the associated `enum` entry.

Reply:
Done